Implement aggregations on aggregate metrics

csoulios opened the pull request:
Following the implementation of the `aggregate_metric_double` field mapper(#49830) we are implementing the Min, Max, ValueCount, Sum and Average aggregations on aggregate metrics.

The code builds on the excellent work done for #42949 and uses the extensible ValuesSources infrastructure to wire up common metric aggregation on the `aggregate_metric_double` field type. 

This PR is part of the rollups v2 refactoring as described in meta issue #42720

csoulios committed:
Implemented aggregate_metric field type for storing pre-computed aggregates

csoulios committed:
Addressing code review comments

csoulios committed:
Moved AggregateMetricFieldMapper to its own module

csoulios committed:
Fix broken doc test

csoulios committed:
Fixes to address code review comments

csoulios committed:
Fixes to address code review comments.

Make parsing of malformed values stricter and
ignore all field metrics even if a single metric
is wrong

csoulios committed:
Fix broken integration tests

csoulios committed:
Added delegate field mappers of NumberFieldType

csoulios committed:
Delegate queries to NumberFieldType fields

csoulios committed:
Style fixes

csoulios committed:
Nit: removed blank line

csoulios committed:
Addressed reviewer comments

csoulios committed:
Fixed NPE issue when "metrics" field is missing

csoulios committed:
Added integration test

csoulios committed:
Override AggregateDoubleMetricFieldMapper methods

isFieldWithinQuery to handle range queries
efficiently,
fielddataBuilder to handle aggregations and sort,
docValueFormat to format double correctly

csoulios committed:
Ensure that a metric field cannot be an array

csoulios committed:
Ensure that merging two fields with different

metrics should be an error.

csoulios committed:
Checkstyle

csoulios committed:
Fix typo

csoulios committed:
Build changes

csoulios committed:
First draft of VS based aggs on aggregate metrics

csoulios committed:
Moved field data related classes to x-pack plugin

csoulios committed:
Removed AggregateDoubleMetricValues classes.

Now AggregateMetricsValuesSource returns directly
a SortedNumericDoubleValues instance which
received by the aggregator

csoulios committed:
Added more tests + code cleanup

csoulios committed:
And more tests

csoulios committed:
Implemented avg agg for aggregate_metric field

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Rollup)

csoulios committed:
Fixed license issues

csoulios committed:
Fix broken tests

csoulios committed:
Implemented min/max aggs on aggregate_metric field

csoulios committed:
Applied "spotless" to mapper-aggregate-metric

csoulios committed:
Added unit tests for AggregateMetricBackedMinAggregator

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/aggregations/metrics/AggregateMetricBackedMaxAggregator.java:
```diff
@@ -0,0 +1,200 @@
+                throw new CollectionTerminatedException();
+            }
+        }
+        if (pointConverter != null) {
```
Future todo (not for this PR): I wonder if there is a way we could extract the BKD optimization in Core so that it can be shared by Core + Pre-aggregated metrics.

But not for today, future project :)

csoulios replied:
For this PR I removed all point reader / bkd optimizations

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/aggregations/metrics/AggregateMetricBackedMaxAggregator.java:
```diff
@@ -0,0 +1,200 @@
+            maxes.fill(0, maxes.size(), Double.NEGATIVE_INFINITY);
+        }
+        this.formatter = config.format();
+        this.pointConverter = getPointReaderOrNull(context, parent, config);
```
I think this won't end up working, or rather, will never return a valid pointConverter and so all the associated BKD logic is not needed.  E.g. `MinAggregator#getPointReaderOrNull()` fetches the field context and then does naive `instanceof` checks to see what kind of field it is (number or date atm).  Which I believe won't match because the field here is actually the pre-aggregated field class.

We'd need some way to tell that function (or create a new one) which metric to get out of the pre-agg field, which is harder since it's in Core.

It's a nice optimization but also pretty niche... I think it would be ok to remove all the point reader / bkd optimization from these aggregators to keep them simple.  We can add them back later.

csoulios replied:
Removed all point reader / bkd optimization

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/aggregations/metrics/AggregateMetricBackedMinAggregator.java:
```diff
@@ -0,0 +1,232 @@
+            mins.fill(0, mins.size(), Double.POSITIVE_INFINITY);
+        }
+        this.format = config.format();
+        this.pointConverter = getPointReaderOrNull(context, parent, config);
```
Ditto here too

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/aggregations/support/AggregateMetricsValuesSourceType.java:
```diff
@@ -0,0 +1,63 @@
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+            final IndexFieldData<?> indexFieldData = fieldContext.indexFieldData();
+
+            if (!(indexFieldData instanceof IndexAggregateDoubleMetricFieldData)) {
```
Tiny nit: `!(indexFieldData ... )` instead of `... == false`

polyfractal reviewed (commented):
Had a look at the aggregator stuff (mostly skimmed over the field mapper changes).  Looks good!  Left a comment about the BKD optimizations for min/max aggs... I think they won't ever do anything based on how the Core pointReader stuff is configured.  We could try to fix... but I think it'd be easier/simpler to just remove all the BKD optimizations for now and add it later.

The only other general comment is regarding the module and if we want this in Rollup or in it's own module.  Not sure the best option tbh.

jimczi commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,689 @@
+        if (other.metrics.explicit()) {
+            if (this.metrics.value() != null
+                && metrics.value().isEmpty() == false
+                && metrics.value().containsAll(other.metrics.value()) == false) {
```
should also disallow removing a metric from an  existing mapping?

jimczi reviewed (commented):
I reviewed the new field mapper and left some minor comments but it's very close.

csoulios committed:
Merge branch 'feature/aggregate-metrics' into prototype/vs-refactor-aggregate-metrics

csoulios committed:
Addressed review comment

csoulios committed:
Added tests for min aggregator

csoulios committed:
Added tests for max aggregator

csoulios committed:
Removed BKD optimizations for min/max aggs

csoulios committed:
Added ValueCount agg backed by aggregate_metric

csoulios committed:
Merge branch 'feature/aggregate-metrics' into prototype/vs-refactor-aggregate-metrics

csoulios committed:
Merged changes from master

csoulios committed:
Fixed bug with partial saving of agg metrics

When ignore_malformed = true and a validation
failed for an agg metric subfield. The agg metric
would be partially saved by the NumberFieldMapper

csoulios committed:
Merge branch 'feature/aggregate-metrics' into prototype/vs-refactor-aggregate-metrics

csoulios committed:
Removed pipeline aggregators to merge with master

csoulios committed:
Fix test

csoulios committed:
Removed multifield and copyTo settings

csoulios committed:
Modified metric subfield separator

csoulios commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -47,6 +66,18 @@
 public class AggregateDoubleMetricFieldMapper extends FieldMapper {
 
     public static final String CONTENT_TYPE = "aggregate_metric_double";
+    public static final String SUBFIELD_SEPARATOR = ".";
```
When I changed the `SUBFIELD_SEPARATOR` from `._` to `.` the yaml test `Test term query on aggregate metric field` started failing because it returned 0 hits. When I change the separator back to `._`, test becomes successful.

I debugged the test but values seem to be stored correctly. @polyfractal @jimczi any idea what I am missing here?

Link to test: https://github.com/elastic/elasticsearch/blob/b9bf2599eec5896d0c2412373231cb3700bc4442/x-pack/plugin/src/test/resources/rest-api-spec/test/aggregate-metrics/10_basic.yml#L57

csoulios replied:
Issue solved. Many thanks to @polyfractal for spotting the problem.

csoulios commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,689 @@
+                                        return new SortedNumericDoubleValues() {
+                                            @Override
+                                            public int docValueCount() {
+                                                return 1;
```
Ensuring that values are not arrays are validated here:
https://github.com/elastic/elasticsearch/blob/b9bf2599eec5896d0c2412373231cb3700bc4442/x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java#L589

This code is tested at https://github.com/elastic/elasticsearch/blob/b9bf2599eec5896d0c2412373231cb3700bc4442/x-pack/plugin/mapper-aggregate-metric/src/test/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapperTests.java#L669

Is there anything else I should implement?

csoulios committed:
Merge branch 'feature/aggregate-metrics' into prototype/vs-refactor-aggregate-metrics

csoulios committed:
Implemented plugin getMinimalSupportedVersion()

csoulios committed:
Fixed broken integration test
Fixed stylechecker error

csoulios committed:
Implemented sortfield for aggregate_metric mapper

csoulios commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,689 @@
+                            XFieldComparatorSource.Nested nested,
+                            boolean reverse
+                        ) {
+                            throw new UnsupportedOperationException("can't sort on the [" + CONTENT_TYPE + "] field");
```
Fixed

csoulios committed:
Added yml tests for aggs on aggregate_metric field

And fixed issues that were discovered in the
process of testing

csoulios commented:
@elasticmachine run elasticsearch-ci/2
@elasticmachine run elasticsearch-ci/bwc

csoulios commented:
@elasticmachine run elasticsearch-ci/default-distro
@elasticmachine run elasticsearch-ci/bwc

csoulios committed:
Added more yml tests

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc

csoulios committed:
Merge branch 'feature/aggregate-metrics' into prototype/vs-refactor-aggregate-metrics

csoulios committed:
Fixed errors because of merge

csoulios committed:
Merge branch 'feature/aggregate-metrics' into prototype/vs-refactor-aggregate-metrics

csoulios committed:
Muted failing tests

csoulios committed:
Merge branch 'prototype/vs-refactor-aggregate-metrics' into value-count-agg-metrics

csoulios committed:
Prepare merge value_count for aggregate metrics

csoulios committed:
Added yml integration test for value_count

on aggregate metric fields

csoulios commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -352,10 +395,118 @@ public Relation isFieldWithinQuery(
+                                            @Override
+                                            public double nextValue() throws IOException {
+                                                long v = values.nextValue();
+                                                if (metric == Metric.value_count) {
+                                                    // Only value_count metrics are encoded as integers
+                                                    return v;
+                                                } else {
+                                                    // All other metrics are encoded as doubles
+                                                    return NumericUtils.sortableLongToDouble(v);
+                                                }
```
When parsing and inserting a `value_count` metric we took the decision to treat as `integer` (see https://github.com/elastic/elasticsearch/blob/32be5aae223a5e876cc4aa1d6ca7890b01cb194c/x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java#L200)

So when we aggregate documents, we must treat `value_count` sub-fields as longs and not as doubles. This part of the code is a bit ugly. Is there any better way to do this?

csoulios committed:
Do no allow merging mappers with missing metrics

csoulios committed:
Merge branch 'feature/aggregate-metrics' into prototype/vs-refactor-aggregate-metrics

csoulios committed:
Merge with master - fixes

csoulios commented:
@elasticmachine run elasticsearch-ci/2
@elasticmachine run elasticsearch-ci/bwc

csoulios committed:
Fixed broken tests

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc

csoulios committed:
Merge branch 'feature/aggregate-metrics' into prototype/vs-refactor-aggregate-metrics

csoulios committed:
Merge branch 'feature/aggregate-metrics' into prototype/vs-refactor-aggregate-metrics

csoulios committed:
Fix build issues from last merge
